docs/proposals/trimming-friendly-registrar-proposal.md

```diff
@@ -0,0 +1,483 @@
+    - all libraries authors will have to publish their code specifically for .NET 9 and any future releases
+      which change the registrar
+  - generating Objective-C code in roslyn source generators is a _hack_ that isn't officially supported
+    - source generators are _synchronous_ and doing many IO operations will slow them down, causing lagging in the editor
+  - its's not yet clear how to correctly generate the `[DynamicDependency]` attributes and if the inverse dependency
+    attributes could be added to .NET
+
+### Notes
+
+- I suggest implementing the change with the Mono.Cecil-based approach since it should be much faster to implement and
+  overall less risky.
+- The C# source generator is a more tempting option and probably more future-proof.
+  - I'm worried it could take a lot longer to implement and there might have problems with third party libraries.
+- This registrar relies on static linking and doesn't support the non-AOT scenario (using `xamarin_registrar_dlsym` to
+  load function pointers to the registrar callbacks).
+    - This is not a problem in the context of NativeAOT but it might mean that the registrar could not completely replace
+      the existing MSR and static registrar for Mono-based (CoreCLR-based?) scenarios and there will still be need
+      to keep the custom linker steps around to build those registrars.
+- There are certainly edge-cases that aren't covered in this document but they should all be possible to implement
+  using the same ideas that are outlined in this document.
+  - This claim needs validation though. Are there some obvious cases which would be hard or obviously inefficient
+    to implement this way?
+- It is unclear what would be the file size and startup or runtime performance implications of these changes.
+  - We will need to generate less managed code but on the other hand we'll generate more Objective-C code.
+  - There will be more transitions between managed and native code. Does it matter too much in the context of NativeAOT
+    though? Will it be worse than performing linear search in the lookup tables that we do today?
+
+---
+
+## Risks and Benefits
+
+### Benefits
+
+- It should be straightforward to modify the existing runtime code and custom linker steps to validate the proposed approach.
+- We would be able to completely drop IL Linker from the build process when publishing with NativeAOT.
+  - The new design will be trimmable using any IL linking tool.
+- There is potential for file size and performance improvements, although this needs validating.
+- It should be possible to generate the code of the registrar using roslyn source generator instead of IL weaving.
+
+### Risks
+
+- We still need the current setup for all the cases which are unsupported by this scenario (inner dev loop, support
+  for existing codebases that rely on the static registrar). The maintenance cost might not decrease, but instead
```

Comment:
I believe it shouldn't be too much of a burden to have (yet) another registrar, especially if we can share at least some of the existing code. Once any new registrars are deemed better/feature complete, we can deprecate/remove any old ones.

Reply:
There are two very distinct parts of the registrar: figure out what needs to be registered, and then doing the actual registration. The first part is currently shared between all registrars (for the dynamic registrar this happens at runtime inside Microsoft.iOS.dll, for the static registrars this happens as a custom linker step, so in _very_ different execution environments). The second part is very different between the dynamic and the static registrars, and this hasn't been much of a problem in the past, so I expect it won't be for a third registrar either. Once again: a good test suite makes this much easier.

Reply:
I see, thanks for the explanation!